### PR TITLE
Remove global variables from core.

### PR DESCRIPTION
**Pull request by jwhui**
This PR removes the use of global variables from the core, with the exception of the API layer (`openthread.cpp`) and `otPlat*`  callbacks (e.g. `mac.cpp`, `timer.cpp`).

The changes are separated into a number of commits.  Each commit seeks to remove global variables from a given source file.  Given the scope of the changes, it might be easier to view each commit separately.

This is related to #506.

### CODE REVIEW

**Comment by mszczodrak** [comment]
LGTM :+1:

File: src/core/common/message.cpp [hunk 1]
```diff
@@ -38,88 +38,78 @@
...
+    VerifyOrExit((message = static_cast<Message *>(NewBuffer())) != NULL, ;);
 
-    return kThreadError_None;
+    memset(message, 0, sizeof(*message));
+    message->SetMessagePool(this);
+    message->SetType(aType);
+    message->SetReserved(aReserved);
+    message->SetLinkSecurityEnabled(true);
+
+    VerifyOrExit(message->SetLength(0) == kThreadError_None, Free(message));
```
Comment (nibanks) [inline src/core/common/message.cpp]: Shouldn't you be setting message to NULL on failure here?

Comment (jwhui) [inline src/core/common/message.cpp]: Good catch.  Will update.

File: src/core/thread/link_quality.hpp [hunk 2]
```diff
@@ -214,23 +221,28 @@ class LinkQualityInfo
...
+void AddNoiseFloor(LinkQualityInfo &aNoiseFloor, int8_t aNoise);
 
 /**
  * This method clears the current average noise floor value.
  *
+ * @param[in]  aNoiseFloor  A reference to the noise state.
+ *
  */
-void ClearNoiseFloorAverage(void);
+void ClearNoiseFloorAverage(LinkQualityInfo &aNoiseFloor);
```
Comment (nibanks) [inline src/core/thread/link_quality.hpp]: Where are all these NoiseFloor functions even used? A quick search of the code indicates no references. As far as I can tell, aNoiseFloor is never anything other than the default value.

Comment (jwhui) [inline src/core/thread/link_quality.hpp]: You are correct that the noise floor estimator is not currently hooked in.  This is an open TODO.

**Comment by nibanks** [comment]
LGTM 👍

File: tests/unit/Makefile.am [hunk 3]
```diff
@@ -129,8 +132,8 @@ test_message_SOURCES         = test_message.cpp
 test_ncp_buffer_LDADD        = $(COMMON_LDADD)
 test_ncp_buffer_SOURCES      = test_ncp_buffer.cpp
 
-test_timer_LDADD             = $(COMMON_LDADD)
-test_timer_SOURCES           = test_timer.cpp
+#test_timer_LDADD             = $(COMMON_LDADD)
+#test_timer_SOURCES           = test_timer.cpp
```
Comment (darconeous) [inline tests/unit/Makefile.am]: Why was this commented out?

Comment (jwhui) [inline tests/unit/Makefile.am]: Oops, that snuck in there.  Just pushed an update to remove this change.

**Comment by darconeous** [comment]
LGTM 👍

**Pull request merged**